File: cpp/znode_demo/sub.cc
#include <iostream>
#include <unistd.h>
#include <trunk/cater/znode.h>
#include <trunk/msgs/hdmap.pb.h>

using namespace trunk;

void hdmap_callback(const msgs::Position& msg) {
    std::cout << "hdmap_callback:" << msg.DebugString() << std::endl;
    std::cout << "---" << std::endl;
}

int main() {
    std::cout << "Hello, Znode Sub!" << std::endl;
    cater::ZNode znode;
    std::string channel = "ipc:///var/log/trunk/haha.ipc";

    znode.channel(channel, 1, cater::ZNode::SUB)
        .sub<msgs::Position>("/haha/hdmap", hdmap_callback)
        .start();

    while (true) {
        znode.spin_once();
    }
    return 0;
}

File: cpp/znode_demo/sub.py
import zmq
from protos.hdmap_pb2 import Position


class Subscriber(object):

    def __init__(self, topic):
        self.topic = topic
        self.data = None
        self.position = Position()
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.SUB)
        self.socket.connect("ipc:///var/log/trunk/haha.ipc")
        self.socket.setsockopt_string(zmq.SUBSCRIBE, "")

    def run(self):
        while True:
            topic, self.data = self.socket.recv_multipart()
            self.position.ParseFromString(self.data)
            print("topic:{} data:{}".format(topic, self.position))


if __name__ == '__main__':
    print("sub topic")
    sub = Subscriber("/haha/hdmap")
    sub.run()
